jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcher.kt

```diff
@@ -0,0 +1,282 @@
+ * @param windowSize     ho many async operations with uncompleted result are allowed in dispatcher
+ * @param closingTimeout max amount of time (in milliseconds) to wait for pending operations during shutdown.
+ *                       If parameter equals 0 then dispatcher will not wait pending operations during closing process.
+ *                       Any negative number will be interpreted as 0.
+ *
+ */
+class SuspendableRateLimitedDispatcher(
+        name: String,
+        rateLimiter: RateLimiter,
+        profiler: Profiler,
+        windowSize: DynamicProperty<Int>,
+        closingTimeout: DynamicProperty<Long>,
+        coroutineScope: CoroutineScope = DispatcherCommonPoolScope
+) : AutoCloseable {
+
+    companion object {
+        private const val QUEUE_SIZE_INDICATOR = "queue_size"
+        private const val ACTIVE_ASYNC_OPERATIONS = "active_async_operations"
+        private const val WINDOW_ACQUIRE_DELAY_MS = 500L
+    }
+
+    private val logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
+
+    private val state = AtomicReference<State>()
+
+    private val rateLimiter: RateLimiter
+    private val deferreds = CopyOnWriteArrayList<Deferred<Any?>>()
+
+    private val name: String
+    private val profiler: Profiler
+
+    /**
+     * Accessed only by single dedicated thread
+     */
+    private var windowSize = 1
+    private val closingTimeout: DynamicProperty<Long>
+
+    private val windowSizeSubscription: PropertySubscription<Int>
+
+    /**
+     * How many async operations executed but their results are not ready yet.
+     */
+    private val activeAsyncOperations = AtomicInteger()
+    private val queueSize = AtomicInteger()
+
+    private val semaphore: AtomicReference<Semaphore> = AtomicReference(Semaphore(1))
+
+    private val coroutineScope: CoroutineScope
+
+    /**
+     * Creates new dispatcher instance
+     */
+    init {
+        this.name = name
+        this.rateLimiter = rateLimiter
+        this.closingTimeout = closingTimeout
+
+        this.profiler = PrefixedProfiler(profiler, "RateLimiterDispatcher.$name.")
+
+        state.set(State.RUNNING)
+
+        this.profiler.attachIndicator(QUEUE_SIZE_INDICATOR) { queueSize.get().toLong() }
+        this.profiler.attachIndicator(ACTIVE_ASYNC_OPERATIONS) { activeAsyncOperations.get().toLong() }
+
+        this.coroutineScope = coroutineScope
+
+        windowSizeSubscription = windowSize.createSubscription().setAndCallListener { oldValue: Int?, newValue: Int ->
+            runBlocking {
+                ChangeWindowSizeCommand(oldValue ?: 0, newValue).apply()
+            }
+        }
+    }
+
+    private fun asyncOperationStarted() {
+        activeAsyncOperations.incrementAndGet()
+    }
+
+    private fun asyncOperationCompleted() {
+        activeAsyncOperations.decrementAndGet()
+        semaphore.get().release()
+    }
+
+    suspend fun <T> compose(supplier: () -> T): Deferred<T> {
```

Comment:
I think such API doesn't comply with coroutines structured concurrency. 

You eighter make fully suspendable function, returning not Job nor Deferred.

Or you make non suspendable function that returnes Job/Deferred.

Here I think out target API is:
```suggestion
    suspend fun <T> decorateSuspend(supplier: suspend () -> T): T {
```

Reply:
API changed